### PR TITLE
Tenant id

### PR DESCRIPTION
**Pull request by bquantump**
https://dev.azure.com/azure-mgmt-ex/DotNET%20Management%20SDK/_boards/board/t/DotNET%20Management%20SDK%20Team/Stories/?workitem=5446

### CODE REVIEW

**Comment by chamons** [comment]
I would like to suggest a different workflow if you don't need the draft PR features (thoughts @pakrym ?)

You can push to a branch on a fork and preview the PR without opening it. Then once it's ready for someone to look at, open the draft or "real" PR.

The "problem" is we're getting review notification spam on each PR commit being pushed, when it isn't obvious that when you want an actual review.

**Comment by pakrym** [comment]
Strangely https://github.blog/changelog/2019-02-14-draft-pull-requests/

states:

>. Draft pull requests suppress notifications to CODEOWNERS reviewers and cannot be merged until they are marked as “Ready for review” in the merge box.

But it seems it doesn't

**Comment by chamons** [comment]
Agreed - very strange (and buggy?).

**Comment by m-nash** [comment]
We don't get the benefit of seeing the CI checks if we don't open draft to a branch named feature/* in this fork

File: src/AutoRest.CSharp/Input/CodeModelPartials.cs [hunk 1]
```diff
@@ -174,6 +175,7 @@ internal partial class Paging
         public Operation? NextLinkOperation { get; set; }
     }
 
+
```
Comment (m-nash) [inline src/AutoRest.CSharp/Input/CodeModelPartials.cs:178]: extra lines

File: src/AutoRest.CSharp/Output/Models/Requests/ProviderToken.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        {
+            noPredecessor = false;
+            hasReferenceSuccessor = false;
+            tokenValue = "";
+            isFullProvider = false;
+            isLastProvider = false;
+            hadSpecialReference = false;
+        }
+    }
+}
```
Comment (m-nash) [inline src/AutoRest.CSharp/Output/Models/Requests/ProviderToken.cs]: missing newline

File: src/AutoRest.CSharp/Output/Models/Requests/ProviderToken.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace AutoRest.CSharp.Output.Models.Requests
+{
+    internal class ProviderToken
+    {
+        internal bool noPredecessor;
```
Comment (m-nash) [inline src/AutoRest.CSharp/Output/Models/Requests/ProviderToken.cs:8]: Make these properties with appropriate get / set and PascalCase

File: src/AutoRest.CSharp/Output/Models/Requests/ProviderToken.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        internal bool hasReferenceSuccessor;
+        internal bool isFullProvider;
+        internal string tokenValue;
+        internal bool hadSpecialReference;
+        internal bool isLastProvider;
+        internal ProviderToken()
+        {
+            noPredecessor = false;
```
Comment (m-nash) [inline src/AutoRest.CSharp/Output/Models/Requests/ProviderToken.cs:16]: no need to initialize if its the default value

File: src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs [hunk 5]
```diff
@@ -305,5 +305,102 @@ private static string ConstructResourceType(string httpRequestUri)
...
+                    var operationMatch = VerifyOperation(token.tokenValue, providerName);
+                    foundNonTenant = !foundNonTenant ? token.isFullProvider && !token.noPredecessor && operationMatch : true;
+                    foundTenant = !foundTenant ? token.isFullProvider && token.noPredecessor && operationMatch : true;
+                }
+            }
+            return foundTenant && !foundNonTenant;
+        }
+
+        //Extensions algo will use same tokens ADO #5523
+        public static List<ProviderToken> Tokenize(string path)
```
Comment (m-nash) [inline src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs:328]: I feel like this method and MakeTokens belong to a different class that OutputLibrary can use

**Review by m-nash — CHANGES_REQUESTED** [review]

File: src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs [hunk 6]
```diff
@@ -305,5 +307,66 @@ private static string ConstructResourceType(string httpRequestUri)
             }
             return null;
         }
+
+        public bool IsTenantOnly(OperationGroup operationGroup)
```
Comment (m-nash) [inline src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs:311]: Can we move this to a TenantDetection class?

File: src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs [hunk 7]
```diff
@@ -305,5 +307,66 @@ private static string ConstructResourceType(string httpRequestUri)
...
+                            break;
+                        }
+                    }
+                }
+            }
+            return foundTenant;
+        }
+
+        //Extensions algo will use same tokens ADO #5523
+        public static List<ProviderSegment> GetProviderSegments(string path)
```
Comment (m-nash) [inline src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs:338]: Can we move this to a ProviderSegmentDetection class?

**Review by m-nash — CHANGES_REQUESTED** [review]

File: src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs [hunk 8]
```diff
@@ -305,5 +307,66 @@ private static string ConstructResourceType(string httpRequestUri)
...
+                var tokenLength = nextReference > -1 ? nextReference - offset : pathLen - offset;
+                currentToken.TokenValue = path.Substring(offset, tokenLength);
+                tokens.Add(currentToken);
+                offset = path.IndexOf(Providers, offset + tokenLength);
+                currentToken = new ProviderSegment();
+            } while (offset > -1);
+            return tokens;
+        }
+
+        public bool VerifyOperation(string tokenValue, string resourceType)
```
Comment (m-nash) [inline src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs:367]: weren't we going to remove this?

**Review by m-nash — CHANGES_REQUESTED** [review]

**Review by m-nash — APPROVED** [review]

**Pull request merged**